### DISCUSSION TITLE
Shell/Bash tasks in Windows using the task runner

### DISCUSSION THREAD

**Discussion by Eric-Bonneau** [alirezanet/Husky.Net#48] [Q&A] [answered]
Hi there,

I've been struggling to try to some Shell/Bash tasks to work in Windows.

To start with, I had this configured:

`task-runner.json`
```json
{
   "tasks": [
   ]
}
```

`pre-commit`
```sh
#!/bin/sh

. "$(dirname "$0")/_/husky.sh"

sh --version
```

Wish was working in `Cmd`:

![image](https://user-images.githubusercontent.com/13652737/178473756-318af74d-4c8b-4345-9ecd-bca239e92587.png)

I then tried to see if I could move the `sh --version` part to its own file and configured it like so:

`display-version.sh`
```sh
#!/bin/sh

sh --version
```

`task-runner.json`
```json
{
   "tasks": [
      {
         "name": "display-version",
         "group": "pre-commit",
         "cwd": ".husky",
         "command": "bash",
         "args": [ "-c", "display-version.sh" ]
      }
   ]
}
```

`pre-commit`
```sh
#!/bin/sh

. "$(dirname "$0")/_/husky.sh"

dotnet husky run
```

But when committing in `Cmd`, I'm getting the following error:

![image](https://user-images.githubusercontent.com/13652737/178474127-c8e785f6-f4fb-485f-9f49-2671a250c032.png)

It makes sense since Bash isn't available in a Windows environment. But I'm wondering how it's working when I inline the script within `pre-commit`. All the above configurations work when I run them in Git Bash instead of `Cmd`. 

**Is there any way I can define tasks that will allow them to run like they do when inlined in `pre-commit`?**

An alternative solution I was playing around with was trying to see if I could run my `display-version.sh` script from my Git Bash installation with a command like: `cmd /c "%PROGRAMFILES%\Git\bin\sh.exe" .husky\display-version.sh` which seems to work:

![image](https://user-images.githubusercontent.com/13652737/178475144-a2ae851a-ae3e-49dd-80b0-71213d64228c.png)

When I try to write an equivalent like so:

`task-runner.json`
```json
{
   "tasks": [
      {
         "name": "display-version",
         "group": "pre-commit",
         "cwd": ".husky",
         "command": "cmd",
         "args": [ "/c", "%PROGRAMFILES%\\Git\\bin\\sh.exe", "display-version.sh" ]
      }
   ]
}
```

I get an error message like:

![image](https://user-images.githubusercontent.com/13652737/178475455-e18eea03-33a8-49d1-a3ed-07666457d53b.png)

I've tried many different ways of putting the `%PROGRAMFILES%` in quotes, using escape characters and such and have not been able to get it to work. I'm not sure if this is a bug with the way the tasks in the task runner are defined, or if I'm just doing something really silly.

In any case, I would really like to use tasks and the task runner. But at the moment, I'm stuck having to inline my Shell/Bash tasks within the hooks like `pre-commit`. I suppose another alternative could be to port my script to a `.csx` file, and rely on `dotnet husky exec`. But I want to avoid that for now unless I have to.

Any feedback is appreciated!

Thanks 😀.

**Comment by alirezanet** [comment, marked as answer]
Hi @Eric-Bonneau 

Unfortunately `Process.Start` in Dotnet doesn't support this format (`%PROGRAMFILES%`) that's why you can not pass this value, but for now, you have a few workarounds to access sh.exe: 

1. Adding `C:\Program Files\Git\bin`  to your system environment PATH, this way sh.exe would be available through cmd. (also there is an option for this during git installation)
2. HardCode `C:\Program Files\Git\bin\sh.exe` instead of `%PROGRAMFILES%\...` 
3. Use a CSX file to run your script. e.g 
```csharp
private var shPath = Path.Combine( Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles) , @"git\bin\sh.exe" );
private var startInfo = new ProcessStartInfo(shPath, "display-version.sh");
Process.Start(startInfo);
```
4. Open an issue and wait until I add support for Dotnet SpecialFolders as a variable (could be a cool feature) 😉🤔

**Reply by alirezanet** [reply]
5- If you can somehow find `sh` path using cmd or PowerShell you can use the [custom variables](https://alirezanet.github.io/Husky.Net/guide/task-configuration.html#custom-variables) to get access to it:

**Reply by Eric-Bonneau** [reply]
Regarding your 2nd suggestion, I was unable to do this. I would always get the `'C:\Program' is not recognized...`. I tried many different ways of escaping `"` characters to get Program Files into something like `"Program Files"`, but it wasn't working properly. I can open an issue with more details regarding what I tried; it's possible there's a bug, or I'm just not trying the right things 😋.

I think running the CSX script (your 3rd suggestion) would be the best long term approach. But rather than having the CSX script call the `.sh` script, it would be better to convert the `.sh` script into a CSX script directly.

Our quick fix for now is to drop the task runner altogether, and just inline our `.sh` script directly in `pre-commit`.
